src/main/java/net/sf/jabref/exporter/BibDatabaseWriter.java

```diff
@@ -33,37 +34,50 @@
 import java.util.stream.Collectors;
 
 import net.sf.jabref.BibDatabaseContext;
-import net.sf.jabref.Globals;
 import net.sf.jabref.MetaData;
 import net.sf.jabref.logic.FieldChange;
-import net.sf.jabref.logic.bibtex.BibEntryWriter;
-import net.sf.jabref.logic.bibtex.LatexFieldFormatter;
 import net.sf.jabref.logic.bibtex.comparator.BibtexStringComparator;
 import net.sf.jabref.logic.bibtex.comparator.CrossRefEntryComparator;
 import net.sf.jabref.logic.bibtex.comparator.FieldComparator;
 import net.sf.jabref.logic.bibtex.comparator.FieldComparatorStack;
 import net.sf.jabref.logic.config.SaveOrderConfig;
 import net.sf.jabref.logic.id.IdComparator;
-import net.sf.jabref.logic.util.strings.StringUtil;
 import net.sf.jabref.model.EntryTypes;
 import net.sf.jabref.model.database.BibDatabase;
+import net.sf.jabref.model.database.BibDatabaseMode;
 import net.sf.jabref.model.entry.BibEntry;
 import net.sf.jabref.model.entry.BibtexString;
 import net.sf.jabref.model.entry.CustomEntryType;
 import net.sf.jabref.model.entry.EntryType;
 
-import org.apache.commons.logging.Log;
-import org.apache.commons.logging.LogFactory;
-
-public class BibDatabaseWriter {
+public abstract class BibDatabaseWriter<E extends SaveSession> {
 
     private static final Pattern REFERENCE_PATTERN = Pattern.compile("(#[A-Za-z]+#)"); // Used to detect string references in strings
-    private static final Log LOGGER = LogFactory.getLog(BibDatabaseWriter.class);
-    private static final String STRING_PREFIX = "@String";
-    private static final String COMMENT_PREFIX = "@Comment";
-    private static final String PREAMBLE_PREFIX = "@Preamble";
-    private BibEntry exceptionCause;
-    private boolean isFirstStringInType;
+    private final SaveSessionFactory<E> saveSessionFactory;
+
+    private E session;
+
+    public BibDatabaseWriter(SaveSessionFactory<E> saveSessionFactory) {
+        this.saveSessionFactory = saveSessionFactory;
+    }
+
+    public interface SaveSessionFactory<E extends SaveSession> {
+        E createSaveSession(Charset encoding, Boolean makeBackup) throws SaveException;
+    }
+
+    private static List<FieldChange> applySaveActions(List<BibEntry> toChange, MetaData metaData) {
+        List<FieldChange> changes = new ArrayList<>();
+
+        Optional<FieldFormatterCleanups> saveActions = metaData.getSaveActions();
+        if (saveActions.isPresent()) {
```

Comment:
Optional only make sense if one uses ifPresent or map/flatMap. Otherwise, it does not really help that much, from my perspective, as one still requires an if check and a get, making it even more complicated to use instead of a null check and just using it.